clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -335,6 +335,14 @@ public void poll(long timeout) throws IOException {
         // we use the time at the end of select to ensure that we don't close any connections that
         // have just been processed in pollSelectionKeys
         maybeCloseOldestConnection(endSelect);
+
+        Iterator<String> connectedIt = connected.iterator();
+        while (connectedIt.hasNext()) {
+            if (disconnected.containsKey(connectedIt.next())) {
+                log.warn("Channel {} is marked connected and disconnected at the same time");
+                connectedIt.remove();
```

Comment:
I was debating the same thing. Won't `NetworkClient` keep the node under the `CONNECTING` state though? It seems like either approach involves a change in the contract that could affect users who are not expecting it. It's an internal class though, so we just need to make sure that the affected Kafka code is updated (if necessary).

It would be nice to include a test for this so that we can verify that things truly work under this scenario.